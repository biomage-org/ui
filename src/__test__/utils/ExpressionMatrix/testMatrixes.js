import { SparseMatrix } from 'mathjs';

const getTwoGenesMatrix = () => ({
  order: ['Gzma', 'Lyz2'],
  rawExpression: new SparseMatrix([
    [1, 2],
    [0, 0],
    [0, 5],
  ]),
  truncatedExpression: new SparseMatrix([
    [1, 2],
    [0, 0],
    [0, 3],
  ]),
  stats: {
    Gzma: {
      rawMean: 0.3, rawStdev: 0.4, truncatedMin: 0, truncatedMax: 1,
    },
    Lyz2: {
      rawMean: 2.3, rawStdev: 2.0, truncatedMin: 2, truncatedMax: 3,
    },
  },
});

const getOtherTwoGenesMatrix = () => ({
  order: ['Hba-x', 'Rbp4'],
  rawExpression: new SparseMatrix([
    [5, 2],
    [8, 9],
    [9, 5],
  ]),
  truncatedExpression: new SparseMatrix([
    [1, 2],
    [1, 5],
    [6, 3],
  ]),
  stats: {
    'Hba-x': {
      rawMean: 5, rawStdev: 0.4, truncatedMin: 10, truncatedMax: 15,
    },
    Rbp4: {
      rawMean: 10, rawStdev: 1.9, truncatedMin: 0, truncatedMax: 6,
    },
  },
});

// eslint-disable-next-line import/prefer-default-export
export { getTwoGenesMatrix, getOtherTwoGenesMatrix };
