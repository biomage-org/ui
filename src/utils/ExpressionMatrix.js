import { SparseMatrix, Index, Range } from 'mathjs';

const getRow = (rowIndex, sparseMatrix) => {
  const [, cellsCount] = sparseMatrix.size();

  return sparseMatrix.subset(Index(rowIndex, Range(0, cellsCount)));
};

const setRow = (rowIndex, newRow, sparseMatrix) => {
  const [, cellsCount] = sparseMatrix.size();

  sparseMatrix.subset(Index(rowIndex, Range(0, cellsCount)), newRow);
};

class ExpressionMatrix {
  constructor() {
    this.lastFreeIndex = 0;

    this.loadedExpressionsIndexes = {};

    this.rawGeneExpressions = new SparseMatrix();
    this.truncatedGeneExpressions = new SparseMatrix();
    // this.zscore = new SparseMatrix();
  }

  getRawExpression(geneSymbol) {
    const geneIndex = this.getIndexFor(geneSymbol);
    const a = getRow(geneIndex, this.rawGeneExpressions).toValue()[0];

    console.log('aDebug');
    console.log(a);
    return a;
  }

  /**
   *
   * @param {*} newGeneSymbols A row with the gene symbols corresponding
   * to each row in the geneExpressions (in the same order)
   * @param {*} newRawGeneExpression A mathjs SparseMatrix with the
   *  raw gene expressions for each of the genes
   * @param {*} newTruncatedGeneExpression A mathjs SparseMatrix with the
   *  raw gene expressions for each of the genes
   */
  pushGeneExpression(newGeneSymbols, newRawGeneExpression, newTruncatedGeneExpression) {
    newGeneSymbols.forEach((geneSymbol, index) => {
      // Get new gene expression
      const newRawGeneExpressionRow = getRow(index, newRawGeneExpression);
      const newTruncatedGeneExpressionRow = getRow(index, newTruncatedGeneExpression);

      // And store it in the matrix
      const geneIndex = this.getIndexFor(geneSymbol);
      setRow(geneIndex, newRawGeneExpressionRow, this.rawGeneExpressions);
      setRow(geneIndex, newTruncatedGeneExpressionRow, this.truncatedGeneExpressions);
    });
  }

  /**
   * If the gene already has an assigned index, it returns it.
   * If it doesn't, it generates a new one for it
   *
   * @param {*} geneSymbol The symbol of the gene
   * @returns The index of the gene inside the raw and truncated matrixes
   */
  getIndexFor(geneSymbol) {
    // If not loaded, assign an index to it
    if (!this.loadedExpressionsIndexes[geneSymbol]) {
      this.loadedExpressionsIndexes[geneSymbol] = this.lastFreeIndex;

      // This index is now assigned, so move it one step
      this.lastFreeIndex += 1;
    }

    return this.loadedExpressionsIndexes[geneSymbol];
  }
}

export default ExpressionMatrix;
