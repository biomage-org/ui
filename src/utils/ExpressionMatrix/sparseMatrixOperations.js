/**
 * Mutates sparseMatrix1
 * Appends the columns of sparseMatrix2 into sparseMatrix1.
 * This method works using the SparseMatrix's of mathjs,
 * which use the structure Compressed Column Matrix,
 * it is efficient at adding/accessing new columns
 *
 * @param {*} sparseMatrix1 Matrix to receive the new columns
 * @param {*} sparseMatrix2 Matrix with the columns to append
 */
// const appendMatrix = (sparseMatrix1, sparseMatrix2) => {
const appendMatrix = () => {
  // sparseMatrix2;
};

// {
//   _index: []
//   _ptr: []
//   _size: []
//   _values: []
// }

// eslint-disable-next-line import/prefer-default-export
export { appendMatrix };
