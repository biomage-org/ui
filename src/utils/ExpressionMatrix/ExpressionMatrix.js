import _ from 'lodash';
import { SparseMatrix } from 'mathjs';

import { appendMatrix, getColumn } from 'utils/ExpressionMatrix/sparseMatrixOperations';

// Commented out pending decision on whether to calculate zScore in the UI or not
// const calculateZScore = (expressionsRow, { rawMean: mean, rawStdev: stdev }) => {
//   const [, cellsCount] = expressionsRow.size();

//   const zScoreRow = new SparseMatrix();

//   cellsCount.forEach((cellIndex) => {
//     const index = new Index(0, cellIndex);

//     const expression = expressionsRow.get(index);

//     const zScore = (expression - mean) / stdev;

//     zScoreRow.set(index, zScore);
//   });

//   const expressionsArray = expressionsRow.valueOf()[0];
//   const zScore = expressionsArray.map((expression) => (
//     expression !== null ? (
//       expression - mean) / stdev
//       : null
//   ));
// };

// const calculateZScore = (responseData) => {
//   const dataWithZScore = Object.entries(responseData).reduce((acc, [gene, value]) => {
//     const { mean, stdev, expression } = value.rawExpression;
//     const zScore = expression.map((x) => (x !== null ? ((x - mean) / stdev) : null));

//     acc[gene] = {
//       ...value,
//       zScore,
//     };

//     return acc;
//   }, {});

//   return dataWithZScore;
// };

class ExpressionMatrix {
  constructor() {
    this.rawGeneExpressions = new SparseMatrix();
    this.truncatedGeneExpressions = new SparseMatrix();
    // this.ZScores = new SparseMatrix();
    this.stats = {};

    this.lastFreeIndex = 0;
    this.loadedExpressionsIndexes = {};
  }

  getExpression(geneSymbol, cellIndexes, matrix) {
    const geneIndex = this.getIndexFor(geneSymbol);

    if (_.isNil(geneIndex)) return undefined;

    const result = getColumn(geneIndex, matrix, cellIndexes);

    // If it's a single number wrap in an array
    if (typeof result === 'number') return [result];
    // If its a matrix transform it to an array
    return result.valueOf().flat();
  }

  getRawExpression(geneSymbol, cellIndexes = undefined) {
    return this.getExpression(geneSymbol, cellIndexes, this.rawGeneExpressions);
  }

  getTruncatedExpression(geneSymbol, cellIndexes = undefined) {
    return this.getExpression(geneSymbol, cellIndexes, this.truncatedGeneExpressions);
  }

  getStats(geneSymbol) {
    return this.stats[geneSymbol];
  }

  geneIsLoaded(geneSymbol) {
    return !_.isNil(this.loadedExpressionsIndexes[geneSymbol]);
  }

  genesAreLoaded(geneSymbols) {
    return geneSymbols.every((geneSymbol) => this.geneIsLoaded(geneSymbol));
  }

  getStoredGenes() {
    return Object.keys(this.loadedExpressionsIndexes);
  }

  setGeneExpression(newGeneSymbols, newRawGeneExpression, newTruncatedGeneExpression, stats) {
    this.rawGeneExpressions = newRawGeneExpression;
    this.truncatedGeneExpressions = newTruncatedGeneExpression;
    this.stats = stats;

    this.loadedExpressionsIndexes = newGeneSymbols.reduce((acum, currentSymbol, index) => {
      // eslint-disable-next-line no-param-reassign
      acum[currentSymbol] = index;
      return acum;
    }, {});

    this.lastFreeIndex = newGeneSymbols.length;
  }

  /**
   *
   * @param {*} newGeneSymbols A row with the gene symbols corresponding
   * to each row in the geneExpressions (in the same order)
   * @param {*} newRawGeneExpression A mathjs SparseMatrix with the
   *  raw gene expressions for each of the genes
   * @param {*} newTruncatedGeneExpression A mathjs SparseMatrix with the
   *  raw gene expressions for each of the genes
   * @param {*} newStats An object which with the stats for each gene's expression
   * Each key is a gene symbol,
   * Each value has this shape: {rawMean, rawStdev, truncatedMin, truncatedMax}
   */
  pushGeneExpression(newGeneSymbols, newRawGeneExpression, newTruncatedGeneExpression, newStats) {
    const [, genesCount] = this.rawGeneExpressions.size();

    // If the matrix was empty previously we can just replace it with the ones that are being pushed
    if (genesCount === 0) {
      this.setGeneExpression(
        newGeneSymbols,
        newRawGeneExpression,
        newTruncatedGeneExpression,
        newStats,
      );
      return;
    }

    // Append new gene expressions
    appendMatrix(this.rawGeneExpressions, newRawGeneExpression);
    appendMatrix(this.truncatedGeneExpressions, newTruncatedGeneExpression);

    // Add new gene stats
    _.merge(this.stats, newStats);

    // Store indexes for the new genes
    newGeneSymbols.forEach((geneSymbol) => {
      this.generateIndexFor(geneSymbol);
    });
  }

  getIndexFor(geneSymbol) {
    return this.loadedExpressionsIndexes[geneSymbol];
  }

  /**
   * If the gene already has an assigned index, it returns it.
   * If it doesn't, it generates a new one for it
   *
   * @param {*} geneSymbol The symbol of the gene
   * @returns The index of the gene inside the raw and truncated matrixes
   */
  generateIndexFor(geneSymbol) {
    // If not loaded, assign an index to it
    if (_.isNil(this.loadedExpressionsIndexes[geneSymbol])) {
      this.loadedExpressionsIndexes[geneSymbol] = this.lastFreeIndex;

      // This index is now assigned, so move it one step
      this.lastFreeIndex += 1;
    }

    return this.loadedExpressionsIndexes[geneSymbol];
  }
}

export default ExpressionMatrix;
