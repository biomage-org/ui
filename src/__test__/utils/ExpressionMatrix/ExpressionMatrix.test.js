// import { SparseMatrix } from 'mathjs';

import ExpressionMatrix from 'utils/ExpressionMatrix/ExpressionMatrix';
import { getTwoGenesMatrix } from '__test__/utils/ExpressionMatrix/testMatrixes';

describe('ExpressionMatrix', () => {
  let matrix;
  describe('Empty matrix', () => {
    beforeEach(() => {
      matrix = new ExpressionMatrix();
    });

    it('getRawExpression works', () => {
      expect(matrix.getRawExpression('Gzma')).toBeUndefined();
    });

    it('getTruncatedExpression works', () => {
      expect(matrix.getRawExpression('Gzma')).toBeUndefined();
    });

    it('getStats works', () => {
      expect(matrix.getStats('Gzma')).toBeUndefined();
    });

    it('geneIsLoaded works', () => {
      expect(matrix.geneIsLoaded('Gzma')).toEqual(false);
    });

    it('genesAreLoaded works', () => {
      expect(matrix.genesAreLoaded(['Gzma, Lyz2'])).toEqual(false);
    });

    it('getStoredGenes works', () => {
      expect(matrix.getStoredGenes()).toEqual([]);
    });

    it('setGeneExpression works', () => {
      const {
        order, rawExpression, truncatedExpression, stats,
      } = getTwoGenesMatrix();

      matrix.setGeneExpression(order, rawExpression, truncatedExpression, stats);

      expect(matrix).toMatchSnapshot();
    });

    it('pushGeneExpression works', () => {
      const {
        order, rawExpression, truncatedExpression, stats,
      } = getTwoGenesMatrix();

      matrix.pushGeneExpression(order, rawExpression, truncatedExpression, stats);

      expect(matrix).toMatchSnapshot();
    });

    it('generateIndexFor works', () => {
      matrix.generateIndexFor('Gzma');

      expect(matrix).toMatchSnapshot();
    });
  });

  describe('Filled matrix', () => {
    beforeEach(() => {
      matrix = new ExpressionMatrix();
      const {
        order, rawExpression, truncatedExpression, stats,
      } = getTwoGenesMatrix();

      matrix.setGeneExpression(order, rawExpression, truncatedExpression, stats);
    });

    it('getRawExpression on existing gene works', () => {
      expect(matrix.getRawExpression('Lyz2')).toEqual([2, 0, 5]);
    });

    it('getRawExpression on non existing gene works', () => {
      expect(matrix.getRawExpression('NotExist')).toBeUndefined();
    });

    it('getTruncatedExpression works', () => {
      expect(matrix.getTruncatedExpression('Gzma')).toEqual([1, 0, 0]);
    });

    it('getStats works', () => {
      expect(matrix.getStats('Gzma')).toEqual(getTwoGenesMatrix().stats.Gzma);
    });

    it('geneIsLoaded on existing gene returns true', () => {
      expect(matrix.geneIsLoaded('Lyz2')).toEqual(true);
    });

    it('geneIsLoaded on non existing gene returns false', () => {
      expect(matrix.geneIsLoaded('IdontExist')).toEqual(false);
    });

    it('genesAreLoaded on existing genes returns true', () => {
      expect(matrix.genesAreLoaded(['Lyz2', 'Gzma'])).toEqual(true);
    });

    it('genesAreLoaded on a mix returns false', () => {
      expect(matrix.genesAreLoaded(['Lyz2', 'IdontExist', 'Gzma'])).toEqual(false);
    });

    it('getStoredGenes works', () => { });

    it('setGeneExpression works', () => { });

    it('pushGeneExpression works', () => { });

    it('generateIndexFor works', () => { });
  });
});
