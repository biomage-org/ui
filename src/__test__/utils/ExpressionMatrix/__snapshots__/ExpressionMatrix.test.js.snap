// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ExpressionMatrix Empty matrix generateIndexFor works 1`] = `
ExpressionMatrix {
  "geneIndexes": Object {
    "Gzma": 0,
  },
  "lastFreeIndex": 1,
  "rawGeneExpressions": Object {
    "datatype": undefined,
    "index": Array [],
    "mathjs": "SparseMatrix",
    "ptr": Array [
      0,
    ],
    "size": Array [
      0,
      0,
    ],
    "values": Array [],
  },
  "stats": Object {},
  "truncatedGeneExpressions": Object {
    "datatype": undefined,
    "index": Array [],
    "mathjs": "SparseMatrix",
    "ptr": Array [
      0,
    ],
    "size": Array [
      0,
      0,
    ],
    "values": Array [],
  },
}
`;

exports[`ExpressionMatrix Empty matrix pushGeneExpression works 1`] = `
ExpressionMatrix {
  "geneIndexes": Object {
    "Gzma": 0,
    "Lyz2": 1,
  },
  "lastFreeIndex": 2,
  "rawGeneExpressions": Object {
    "datatype": undefined,
    "index": Array [
      0,
      0,
      2,
    ],
    "mathjs": "SparseMatrix",
    "ptr": Array [
      0,
      1,
      3,
    ],
    "size": Array [
      3,
      2,
    ],
    "values": Array [
      1,
      2,
      5,
    ],
  },
  "stats": Object {
    "Gzma": Object {
      "rawMean": 0.3,
      "rawStdev": 0.4,
      "truncatedMax": 1,
      "truncatedMin": 0,
    },
    "Lyz2": Object {
      "rawMean": 2.3,
      "rawStdev": 2,
      "truncatedMax": 3,
      "truncatedMin": 2,
    },
  },
  "truncatedGeneExpressions": Object {
    "datatype": undefined,
    "index": Array [
      0,
      0,
      2,
    ],
    "mathjs": "SparseMatrix",
    "ptr": Array [
      0,
      1,
      3,
    ],
    "size": Array [
      3,
      2,
    ],
    "values": Array [
      1,
      2,
      3,
    ],
  },
}
`;

exports[`ExpressionMatrix Empty matrix setGeneExpression works 1`] = `
ExpressionMatrix {
  "geneIndexes": Object {
    "Gzma": 0,
    "Lyz2": 1,
  },
  "lastFreeIndex": 2,
  "rawGeneExpressions": Object {
    "datatype": undefined,
    "index": Array [
      0,
      0,
      2,
    ],
    "mathjs": "SparseMatrix",
    "ptr": Array [
      0,
      1,
      3,
    ],
    "size": Array [
      3,
      2,
    ],
    "values": Array [
      1,
      2,
      5,
    ],
  },
  "stats": Object {
    "Gzma": Object {
      "rawMean": 0.3,
      "rawStdev": 0.4,
      "truncatedMax": 1,
      "truncatedMin": 0,
    },
    "Lyz2": Object {
      "rawMean": 2.3,
      "rawStdev": 2,
      "truncatedMax": 3,
      "truncatedMin": 2,
    },
  },
  "truncatedGeneExpressions": Object {
    "datatype": undefined,
    "index": Array [
      0,
      0,
      2,
    ],
    "mathjs": "SparseMatrix",
    "ptr": Array [
      0,
      1,
      3,
    ],
    "size": Array [
      3,
      2,
    ],
    "values": Array [
      1,
      2,
      3,
    ],
  },
}
`;

exports[`ExpressionMatrix Filled matrix generateIndexFor works 1`] = `
ExpressionMatrix {
  "geneIndexes": Object {
    "Gzma": 0,
    "Hba-x": 2,
    "Lyz2": 1,
  },
  "lastFreeIndex": 3,
  "rawGeneExpressions": Object {
    "datatype": undefined,
    "index": Array [
      0,
      0,
      2,
    ],
    "mathjs": "SparseMatrix",
    "ptr": Array [
      0,
      1,
      3,
    ],
    "size": Array [
      3,
      2,
    ],
    "values": Array [
      1,
      2,
      5,
    ],
  },
  "stats": Object {
    "Gzma": Object {
      "rawMean": 0.3,
      "rawStdev": 0.4,
      "truncatedMax": 1,
      "truncatedMin": 0,
    },
    "Lyz2": Object {
      "rawMean": 2.3,
      "rawStdev": 2,
      "truncatedMax": 3,
      "truncatedMin": 2,
    },
  },
  "truncatedGeneExpressions": Object {
    "datatype": undefined,
    "index": Array [
      0,
      0,
      2,
    ],
    "mathjs": "SparseMatrix",
    "ptr": Array [
      0,
      1,
      3,
    ],
    "size": Array [
      3,
      2,
    ],
    "values": Array [
      1,
      2,
      3,
    ],
  },
}
`;

exports[`ExpressionMatrix Filled matrix pushGeneExpression adds new data keeping previous 1`] = `
ExpressionMatrix {
  "geneIndexes": Object {
    "Gzma": 0,
    "Hba-x": 2,
    "Lyz2": 1,
    "Rbp4": 3,
  },
  "lastFreeIndex": 4,
  "rawGeneExpressions": Object {
    "datatype": undefined,
    "index": Array [
      0,
      0,
      2,
      0,
      1,
      2,
      0,
      1,
      2,
    ],
    "mathjs": "SparseMatrix",
    "ptr": Array [
      0,
      1,
      3,
      12,
      15,
    ],
    "size": Array [
      3,
      4,
    ],
    "values": Array [
      1,
      2,
      5,
      5,
      8,
      9,
      2,
      9,
      5,
    ],
  },
  "stats": Object {
    "Gzma": Object {
      "rawMean": 0.3,
      "rawStdev": 0.4,
      "truncatedMax": 1,
      "truncatedMin": 0,
    },
    "Hba-x": Object {
      "rawMean": 5,
      "rawStdev": 0.4,
      "truncatedMax": 15,
      "truncatedMin": 10,
    },
    "Lyz2": Object {
      "rawMean": 2.3,
      "rawStdev": 2,
      "truncatedMax": 3,
      "truncatedMin": 2,
    },
    "Rbp4": Object {
      "rawMean": 10,
      "rawStdev": 1.9,
      "truncatedMax": 6,
      "truncatedMin": 0,
    },
  },
  "truncatedGeneExpressions": Object {
    "datatype": undefined,
    "index": Array [
      0,
      0,
      2,
      0,
      1,
      2,
      0,
      1,
      2,
    ],
    "mathjs": "SparseMatrix",
    "ptr": Array [
      0,
      1,
      3,
      12,
      15,
    ],
    "size": Array [
      3,
      4,
    ],
    "values": Array [
      1,
      2,
      3,
      1,
      1,
      6,
      2,
      5,
      3,
    ],
  },
}
`;

exports[`ExpressionMatrix Filled matrix setGeneExpression replaces existing data 1`] = `
ExpressionMatrix {
  "geneIndexes": Object {
    "Hba-x": 0,
    "Rbp4": 1,
  },
  "lastFreeIndex": 2,
  "rawGeneExpressions": Object {
    "datatype": undefined,
    "index": Array [
      0,
      1,
      2,
      0,
      1,
      2,
    ],
    "mathjs": "SparseMatrix",
    "ptr": Array [
      0,
      3,
      6,
    ],
    "size": Array [
      3,
      2,
    ],
    "values": Array [
      5,
      8,
      9,
      2,
      9,
      5,
    ],
  },
  "stats": Object {
    "Hba-x": Object {
      "rawMean": 5,
      "rawStdev": 0.4,
      "truncatedMax": 15,
      "truncatedMin": 10,
    },
    "Rbp4": Object {
      "rawMean": 10,
      "rawStdev": 1.9,
      "truncatedMax": 6,
      "truncatedMin": 0,
    },
  },
  "truncatedGeneExpressions": Object {
    "datatype": undefined,
    "index": Array [
      0,
      1,
      2,
      0,
      1,
      2,
    ],
    "mathjs": "SparseMatrix",
    "ptr": Array [
      0,
      3,
      6,
    ],
    "size": Array [
      3,
      2,
    ],
    "values": Array [
      1,
      1,
      6,
      2,
      5,
      3,
    ],
  },
}
`;
